### PR TITLE
Add python api for optimize

### PR DESCRIPTION
**Pull request by burakyilmaz321**
- [x] Unit tests
- [x] Documentation

## Description

Adds optimize command to the Python API.

- New class in ``tables.py``:  ``DeltaOptimizeBuilder``
- New method: ``DeltaTable.optimize``

Resolves https://github.com/delta-io/delta/issues/1080

## How was this patch tested?

Added new unit tests to ``test_deltatable.py``:

- ``test_optimize``
- ``test_optimize_w_partition_filter``

## Does this PR introduce _any_ user-facing changes?

No

### CODE REVIEW

**Comment by vkorukanti** [comment]
@burakyilmaz321 Thank you for the PR. This is very good! For tests, you can look at the the examples [here](https://github.com/delta-io/delta/blob/master/python/delta/tests/test_deltatable.py#L128).

**Comment by burakyilmaz321** [comment]
@vkorukanti Added unit tests and documentation. Since the Scala API is not added yet, tests are failing here. But, I merged https://github.com/delta-io/delta/pull/961 with my changes in [another branch](https://github.com/burakyilmaz321/delta/tree/python-api-optimize-test), just to see that unit tests are passing. You can see [here](https://github.com/burakyilmaz321/delta/pull/2) the unit test checks are successful.

File: python/delta/tests/test_deltatable.py [hunk 1]
```diff
@@ -850,6 +850,54 @@ def runRestoreToVersion() -> None:
         self.__intercept(runRestoreToVersion,
                          "version needs to be an int but got '<class 'str'>'")
 
+    def test_optimize(self) -> None:
```
Comment (vkorukanti) [inline python/delta/tests/test_deltatable.py:853]: can you also add a negative test which provides non-partition column filter?

Comment (burakyilmaz321) [reply python/delta/tests/test_deltatable.py:853]: @vkorukanti I added tests for non-partition columns.

I also renamed ``partitionFilter`` method as ``where`` as you pointed [here](https://github.com/delta-io/delta/pull/961#discussion_r861114399).

File: python/delta/tables.py [hunk 2]
```diff
@@ -581,6 +581,14 @@ def restoreToTimestamp(self, timestamp: str) -> DataFrame:
             getattr(self._spark, "_wrapped", self._spark)  # type: ignore[attr-defined]
         )
 
+    @since(1.3)  # type: ignore[arg-type]
+    def optimize(self) -> "DeltaOptimizeBuilder":
+        """
+        Optimize the data layout of the table.
```
Comment (vkorukanti) [inline python/delta/tables.py:587]: You can add more details here similar to Scala API: https://github.com/delta-io/delta/blob/master/core/src/main/scala/io/delta/tables/DeltaTable.scala#L186

File: python/delta/tests/test_deltatable.py [hunk 3]
```diff
@@ -850,6 +850,72 @@ def runRestoreToVersion() -> None:
...
+        # write an unoptimized delta table
+        df = self.spark.createDataFrame([("a", 1), ("a", 2)], ["key", "value"]).repartition(1)
```
Comment (vkorukanti) [inline python/delta/tests/test_deltatable.py:855]: can we use the utility methods `__writeDeltaTable`, `__writeAsTable` or `__overwriteDeltaTable` here? Is the number of files hard to control without the repartition for test verification?

File: python/delta/tests/test_deltatable.py [hunk 4]
```diff
@@ -850,6 +850,72 @@ def runRestoreToVersion() -> None:
...
+
+        # assertions
+        self.assertTrue(isinstance(optimizer, DeltaOptimizeBuilder))
+        self.assertTrue(isinstance(res, DataFrame))
+        self.assertEqual(1, res.first().metrics.numFilesAdded)
+        self.assertEqual(3, res.first().metrics.numFilesRemoved)
+        self.assertDictEqual({'predicate': '["true"]'}, op_params)
+
+        # test non-partition column
+        rgx_msg = (
```
Comment (vkorukanti) [inline python/delta/tests/test_deltatable.py:878]: does `__intercept` method help here?

**Review by vkorukanti — CHANGES_REQUESTED** [review]
Thank you @burakyilmaz321 for the update. Few minor comments. The Scala API is merged to master branch.

File: python/delta/tables.py (hunk 2 above)
Comment (burakyilmaz321) [reply python/delta/tables.py]: Added more details to the docstrings. I also rebased to the master.

File: python/delta/tests/test_deltatable.py (hunk 4 above)
Comment (burakyilmaz321) [reply python/delta/tests/test_deltatable.py]: done

File: python/delta/tests/test_deltatable.py (hunk 3 above)
Comment (burakyilmaz321) [reply python/delta/tests/test_deltatable.py:855]: @vkorukanti I checked these but I think I need something that writes unoptimized delta tables. I want to create a delta table with multiple parquet files, then I can execute compaction on them. I used ``repartition`` to control the number of files written, and I used ``partitionBy`` to test the partitioned table case.

I doubt ``__writeDeltaTable``, ``__writeAsTable`` or `` __overwriteDeltaTable`` helps when creating such setting.

Do you have any suggestions? I can remove "numFilesAdded" and "numFilesRemoved" verifications from the test if it is not necessary. I can use ``__writeDeltaTable`` that way.

Or I can add "mode", "repartition" and "partitioned" parameters to ``__writeDeltaTable``. So I can use it as:

```python
self.__writeDeltaTable([("a", 1), ("a", 2)], repartition=1)
self.__writeDeltaTable([("a", 3), ("a", 4)], repartition=1, mode="append")
self.__writeDeltaTable([("b", 1), ("b", 2)], repartition=1, mode="append")

# and
self.__writeDeltaTable([("a", 1), ("a", 2)], repartition=1, partitioned=True)
self.__writeDeltaTable([("a", 3), ("a", 4)], repartition=1, partitioned=True, mode="append")
self.__writeDeltaTable([("b", 1), ("b", 2)], repartition=1, partitioned=True, mode="append")

# where
def __writeDeltaTable(self, datalist, repartition=None, partitioned=False, mode="error"):
    df = self.spark.createDataFrame(datalist, ["key", "value"])
    if repartition:
        df = df.repartition(repartition)
    writer = df.write
    if partitioned:
        writer = writer.partitionBy("key")
    writer.format("delta").save(self.tempFile, mode=mode)
```

Let me what you think.

Comment (vkorukanti) [reply python/delta/tests/test_deltatable.py:855]: Thanks for checking. I think the current method is fine. In future if there are more tests that need such utility methods, we can add them.

**Review by vkorukanti — APPROVED** [review]
LGTM. Thanks for adding these APIs.

**Comment by vkorukanti** [comment]
@burakyilmaz321 The python test failures on `master` branch are fixed now. Could you please rebase this PR? Thank you!

**Comment by burakyilmaz321** [comment]
> @burakyilmaz321 The python test failures on `master` branch are fixed now. Could you please rebase this PR? Thank you!

Sure, I will do that as soon as I reach my laptop later today.

**Pull request closed without merging**